pkg/http/server.go

```diff
@@ -123,9 +125,14 @@ func (srv *Server) Shutdown() error {
 		return errors.New("http server already in shutdown")
 	}
 
+	if atomic.LoadInt32(&srv.upAndRunning) <= 0 {
+		return nil
+	}
+
 	// Close underneath HTTP listener.
 	srv.listenerMutex.Lock()
 	err := srv.listener.Close()
```

Comment:
I see now @ebozduman  checking for listener as nil is fine if we are not resetting or reading the value elsewhere for this fix.

The refactor discussions requires lot more work we can defer that discussion for later date.